python/templates/macros/iterator.jinja2

```diff
@@ -3,27 +3,39 @@
 {% set ptr_init = 'podio::utils::MaybeSharedPtr<' + class.bare_type +'Obj>{nullptr}' %}
 class {{ iterator_type }} {
 public:
+  using value_type = {{ class.bare_type }};
+  using difference_type = ptrdiff_t;
+  using reference = {{ prefix }}{{ class.bare_type }};
+  using pointer = {{ prefix }}{{ class.bare_type }}*;
+  using iterator_category = std::input_iterator_tag;
+  using iterator_concept = std::input_iterator_tag;
+
   {{ iterator_type }}(size_t index, const {{ class.bare_type }}ObjPointerContainer* collection) : m_index(index), m_object({{ ptr_init }}), m_collection(collection) {}
+  {{ iterator_type }}() = default;
 
-  {{ iterator_type }}(const {{ iterator_type }}&) = delete;
-  {{ iterator_type }}& operator=(const {{ iterator_type }}&) = delete;
+  {{ iterator_type }}(const {{ iterator_type }}&) = default;
+  {{ iterator_type }}({{ iterator_type }}&&) = default;
+  {{ iterator_type }}& operator=(const {{ iterator_type }}&) = default;
+  {{ iterator_type }}& operator=({{iterator_type}}&&) = default;
+  ~{{ iterator_type }}() = default;
 
   bool operator!=(const {{ iterator_type}}& x) const {
-    return m_index != x.m_index; // TODO: may not be complete
+    return m_index != x.m_index;
   }
 
   bool operator==(const {{ iterator_type }}& x) const {
-    return m_index ==  x.m_index; // TODO: may not be complete
+    return m_index ==  x.m_index;
   }
 
-  {{ prefix }}{{ class.bare_type }} operator*();
-  {{ prefix }}{{ class.bare_type }}* operator->();
+  reference operator*() const;
+  pointer operator->();
   {{ iterator_type }}& operator++();
+  {{ iterator_type }} operator++(int);
 
 private:
-  size_t m_index;
-  {{ prefix }}{{ class.bare_type }} m_object;
-  const {{ class.bare_type }}ObjPointerContainer* m_collection;
+  size_t m_index = 0;
+  {{ prefix }}{{ class.bare_type }} m_object { {{ ptr_init }} };
+  const {{ class.bare_type }}ObjPointerContainer* m_collection = nullptr;
```

Comment:
```suggestion
  size_t m_index{0};
  {{ prefix }}{{ class.bare_type }} m_object { {{ ptr_init }} };
  const {{ class.bare_type }}ObjPointerContainer* m_collection{nullptr};
```
To stay consistent with how we do it throughout the rest of podio and the generated EDMs.